src/extensions/Relude_Extensions_BoundedEnum.re

```diff
@@ -27,4 +27,185 @@ module BoundedEnumExtensions = (E: Relude_Interface.BOUNDED_ENUM) => {
+            let rec find = (arr, i, limit) =>
+              if (i < limit) {
+                let (e__, a__) = Belt.Array.getUnsafe(arr, i);
+                if (eqA(a__, a)) {
+                  Some(e__);
+                } else {
+                  find(arr, succ(i), limit);
+                };
+              } else {
+                None;
+              };
+            find(arr, 0, len);
+          };
+          return;
+        | Some(e) =>
+          // Continue building the associative list:
+          loop([(e, eToA(e)), ...lst], getNextE(e), getNextE)
+        };
+      };
+
+      // Build associative list and pass a lookup function to the caller:
+      loop([], Some(E.bottom), E.succ);
+    };
+
+  /**
+   [BoundedEnumExtensions.inverseMapEq] takes a first-class [EQ] module for
+   type [a], along with a function of type [t => a] where each [a] is a
+   unique value, and returns a new function of type [a => option(t)].
+
+   This is useful for generating parser utilities, reverse-lookup functions,
+   etc.
+
+   Note: since there is no way to determine the ordering of [a] without
+   a comparator function, any call to the returned function must iterate
+   through the entire list of possible values until it finds a match or
+   reaches the end of the list. Therefore, the worst-case time complexity
+   of the returned function is **linear**, and failed lookups will always
+   be worst-case.
+
+   See {!val:inverseMapToOrd} for a similar function that accepts a
+   comparator for [a], and returns a reverse-lookup function with a
+   worst-case running time of O(log(n)).
+
+   Running time for staging: O(n)
+   Running time of the returned lookup function: O(n)
+   */
+  let inverseMapEq:
+    type a.
+      (~eqA: (module BsBastet.Interface.EQ with type t = a), E.t => a, a) =>
+      option(E.t) =
+    (~eqA, eToA) => {
+      let (module EqA) = (eqA: (module BsBastet.Interface.EQ with type t = a));
+      inverseMapEqBy(EqA.eq, eToA);
+    };
+
+  /**
+   [BoundedEnumExtensions.inverseMapOrdBy] takes a comparator function of
+   type [(a, a) => Relude.Ordering.t], along with a mapping function of type
+   [t => a], where each [a] is a unique value, and returns a new function
+   of type [a => option(t)].
+
+   [inverseMapOrdBy] performs the same operations as [inverseMapOrd], but
+   it accepts a comparator function for [a], instead of a first-class
+   [Ord] module.
+
+   Running time for staging: O(n)
+   Running time for returned lookup function: O(log(n))
+   */
+  let inverseMapOrdBy:
+    type a.
+      ((a, a) => BsBastet.Interface.ordering, E.t => a, a) => option(E.t) =
+    (cmp, eToA) => {
+      // Generate the Map data structure:
+      let (module M) = (
+        (module
+         Map.Make({
+           type t = a;
+           let compare = (a, b) =>
+             switch (cmp(a, b)) {
+             | `equal_to => 0
+             | `less_than => (-1)
+             | `greater_than => 1
+             };
```

Comment:
Right, I remember wanting to use that function, but then running into cyclic dependency issues, since `Relude_Ordering` includes a functor for `BoundedEnumExtensions`. Hence why I even used the raw `BsBastet.Interface.ordering` type instead of Relude's own type alias. I'm open to alternatives, though.

Reply:
Ah, good point... in that case, it's probably not worth worrying about